linera-core/src/client/mod.rs

```diff
@@ -2469,60 +2356,131 @@ where
                         // finalized the block.
-                        let timestamp = manager.round_timeout.ok_or_else(|| {
+                        let timestamp = info.manager.round_timeout.ok_or_else(|| {
                             ChainClientError::BlockProposalError(
                                 "Cannot propose in the current round.",
                             )
                         })?;
                         return Ok(ClientOutcome::WaitForTimeout(RoundTimeout {
                             timestamp,
-                            current_round: manager.current_round,
+                            current_round: info.manager.current_round,
                             next_block_height: info.next_block_height,
                         }));
                     }
                     Err(error) => return Err(error),
                 }
+            } else {
+                // Otherwise we have to re-propose the highest validated block.
+                certificate.executed_block().clone()
             }
-        }
+        } else {
+            let block = if let Some(proposal) = info
+                .manager
+                .requested_proposed
+                .as_ref()
+                .filter(|proposal| proposal.content.round.is_fast())
+            {
+                // The fast block counts as "locked", too. If there is one, re-propose that.
+                proposal.content.block.clone()
+            } else if let Some(block) = self.state().pending_block() {
+                block.clone() // Otherwise we are free to propose our own pending block.
+            } else {
+                return Ok(ClientOutcome::Committed(None)); // Nothing to do.
+            };
+            self.stage_block_execution(block).await?.0
+        };
 
-        // The block we want to propose is either the highest validated, or our pending one.
-        let Some(block) = manager
-            .highest_validated_block()
-            .cloned()
-            .or_else(|| self.state().pending_block().clone())
+        let Some(round) =
+            Self::round_for_new_proposal(&info.manager, &identity, Some(&executed_block.block))
         else {
-            return Ok(ClientOutcome::Committed(None)); // Nothing to propose.
+            let timeout = info.round_timeout().ok_or_else(|| {
+                ChainClientError::BlockProposalError("Cannot propose in the current round.")
+            })?;
+            return Ok(ClientOutcome::WaitForTimeout(timeout));
         };
 
-        let round = match Self::round_for_new_proposal(&info, &identity, Some(&block))? {
-            Either::Left(round) => round,
-            Either::Right(timeout) => return Ok(ClientOutcome::WaitForTimeout(timeout)),
+        let block = executed_block.block.clone();
+        let hashed_value = if round.is_fast() {
+            HashedCertificateValue::new_confirmed(executed_block)
+        } else {
+            HashedCertificateValue::new_validated(executed_block)
```

Comment:
If the `executed_block` here comes from `info.manager.requested_proposed` then we must have already created certificates for it earlier, hadn't we?

Reply:
At least we must have executed the block, yes. But we didn't store it, if we just handled the proposal in the local node.

We should, of course, but that's also going to be a bigger change that I'd like to leave for a later PR.

Reply:
The only place where `proposed` is set is in [create_vote](https://github.com/linera-io/linera-protocol/blob/d5da9e70706f3b919aab731caea5433710c74aaf/linera-chain/src/manager.rs#L417) method, which is followed immediately by creating the vote over the hashed value and insert into cache (and `self.save()`). What do you mean by:

> But we didn't store it, if we just handled the proposal in the local node.

Reply:
Yes, but in the client we don't have a key pair so we don't create the vote.

I agree, we should store the hashed value anyway, but I'd really like to not add that to this PR, too.

Reply:
There are already issues related to this: https://github.com/linera-io/linera-protocol/issues/1401, https://github.com/linera-io/linera-protocol/issues/2270, https://github.com/linera-io/linera-protocol/issues/2912